From 31f639f850a7f28e0d46fb1c046b2fb2cd3d8c74 Mon Sep 17 00:00:00 2001
From: Matthew Schwartz <matthew.schwartz@linux.dev>
Date: Thu, 2 Jul 2026 12:06:22 -0700
Subject: [PATCH 140/188] PCI: Generalize pci_mask_replay_timer_timeout()
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Pass the device to mask as an argument so the helper can also cover
devices that report Replay Timer Timeout errors themselves rather than
through their upstream Port. No functional change intended.

Signed-off-by: Matthew Schwartz <matthew.schwartz@linux.dev>
Signed-off-by: Nícolas F. R. A. Prado <nfraprado@collabora.com>
---
 drivers/pci/quirks.c | 21 +++++++++++++--------
 1 file changed, 13 insertions(+), 8 deletions(-)

diff --git a/drivers/pci/quirks.c b/drivers/pci/quirks.c
index b09f27f78..e3fa06f0d 100644
--- a/drivers/pci/quirks.c
+++ b/drivers/pci/quirks.c
@@ -6397,21 +6397,26 @@ static void pci_fixup_d3cold_delay_1sec(struct pci_dev *pdev)
 DECLARE_PCI_FIXUP_FINAL(0x5555, 0x0004, pci_fixup_d3cold_delay_1sec);
 
 #ifdef CONFIG_PCIEAER
-static void pci_mask_replay_timer_timeout(struct pci_dev *pdev)
+static void pci_mask_replay_timer_timeout(struct pci_dev *pdev, struct pci_dev *reporter)
 {
-	struct pci_dev *parent = pci_upstream_bridge(pdev);
 	u32 val;
 
-	if (!parent || !parent->aer_cap)
+	if (!reporter || !reporter->aer_cap)
 		return;
 
-	pci_info(parent, "mask Replay Timer Timeout Correctable Errors due to %s hardware defect",
+	pci_info(reporter, "mask Replay Timer Timeout Correctable Errors due to %s hardware defect",
 		 pci_name(pdev));
 
-	pci_read_config_dword(parent, parent->aer_cap + PCI_ERR_COR_MASK, &val);
+	pci_read_config_dword(reporter, reporter->aer_cap + PCI_ERR_COR_MASK, &val);
 	val |= PCI_ERR_COR_REP_TIMER;
-	pci_write_config_dword(parent, parent->aer_cap + PCI_ERR_COR_MASK, val);
+	pci_write_config_dword(reporter, reporter->aer_cap + PCI_ERR_COR_MASK, val);
 }
-DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_GLI, 0x9750, pci_mask_replay_timer_timeout);
-DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_GLI, 0x9755, pci_mask_replay_timer_timeout);
+
+/* The Port leading to the GL975x reports the errors. */
+static void quirk_gl975x_mask_replay_timer_timeout(struct pci_dev *pdev)
+{
+	pci_mask_replay_timer_timeout(pdev, pci_upstream_bridge(pdev));
+}
+DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_GLI, 0x9750, quirk_gl975x_mask_replay_timer_timeout);
+DECLARE_PCI_FIXUP_FINAL(PCI_VENDOR_ID_GLI, 0x9755, quirk_gl975x_mask_replay_timer_timeout);
 #endif
-- 
2.55.0

